src/clp_ffi_js/ir/IrStreamReader.hpp

```diff
@@ -0,0 +1,113 @@
+#ifndef CLP_FFI_JS_IR_IR_STREAM_READER_HPP
+#define CLP_FFI_JS_IR_IR_STREAM_READER_HPP
+
+#include <Array.hpp>
+#include <cstddef>
+#include <memory>
+#include <optional>
+#include <streaming_compression/zstd/Decompressor.hpp>
+#include <vector>
+
+#include <clp/ir/types.hpp>
+#include <clp/TimestampPattern.hpp>
+#include <emscripten/val.h>
+
+#include <clp_ffi_js/ir/LogEventWithLevel.hpp>
+#include <clp_ffi_js/ir/StreamReader.hpp>
+#include <clp_ffi_js/ir/StreamReaderDataContext.hpp>
+
+namespace clp_ffi_js::ir {
+using clp::ir::four_byte_encoded_variable_t;
+
+/**
+ * Mapping between an index in the filtered log events collection to an index in the unfiltered
+ * log events collection.
+ */
+using FilteredLogEventsMap = std::optional<std::vector<size_t>>;
+
+/**
+ * Class to deserialize and decode Zstandard-compressed CLP IRv1 streams as well as format decoded
+ * log events.
+ */
+class IrStreamReader : public StreamReader {
+    friend StreamReader;
+
+public:
+    // Destructor
+    ~IrStreamReader() override = default;
+
+    // Disable copy constructor and assignment operator
+    IrStreamReader(IrStreamReader const&) = delete;
+    auto operator=(IrStreamReader const&) -> IrStreamReader& = delete;
+
+    // Define default move constructor
+    IrStreamReader(IrStreamReader&&) = default;
+    // Delete move assignment operator since it's also disabled in `clp::ir::LogEventDeserializer`.
+    auto operator=(IrStreamReader&&) -> IrStreamReader& = delete;
+
+    // Methods
+    /**
+     * @return The number of events buffered.
+     */
+    [[nodiscard]] auto get_num_events_buffered() const -> size_t override;
+
+    /**
+     * @return The filtered log events map.
+     */
+    [[nodiscard]] auto get_filtered_log_event_map() const -> FilteredLogEventMapTsType override;
+
+    /**
+     * Generates a filtered collection from all log events.
+     *
+     * @param log_level_filter Array of selected log levels
+     */
+    void filter_log_events(LogLevelFilterTsType const& log_level_filter) override;
+
+    /**
+     * Deserializes all log events in the stream. After the stream has been exhausted, it will be
```

Comment:
Deallocation seems like an implementation decision, so perhaps we can keep this note in the implementation's docstring but not in the interface's. @junhaoliao, what do you think?

Reply:
Sure

Reply:
Currently, I see you've deleted this docstring. I'm fine keeping it this way, but if @junhaoliao wants to keep the part about deallocating the stream after exhaustion, then we need to add the docstring back in `UnstructuredIrStreamReader`.

Reply:
(just to be sure I understand the discussion - 
* in `StreamReader` which serves an interface, I agree we do not want to mention `After the stream has been exhausted, it will be deallocated.`.
* in `UnstructuredIrStreamReader` which is an implementation of the interface, I prefer that we mention `After the stream has been exhausted, it will be deallocated.` so that we can remind any future developer about the stream data lifecycle.

Reply:
Correct. One way to do this is:

```cpp
    /**
     * Deserializes all log events in the stream. After the stream has been exhausted, it will be
     * deallocated.
     *
     * @return The number of successfully deserialized ("valid") log events.
     */
```

Another way is:

```cpp
    /**
     * See `StreamReader::deserialize_stream`.
     * 
     * After the stream has been exhausted, it will be deallocated.
     *
     * @return See `StreamReader::deserialize_stream`.
     */
```

We usually go with the first approach, but we don't have too many instances of this case, so we could change conventions.

Reply:
I think we can go with the second approach to avoid repetition and docs going out of sync any time in the future.

For the second approach, can we use `@sa` or `@see`? (i'm not too sure if this is how the command is intended to be used: https://www.doxygen.nl/manual/commands.html#cmdsa

Reply:
Didn't know Doxygen had see. Yeah, let's use `@see`.

Reply:
@davemarco, can you make the change?